lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp

```diff
@@ -452,17 +452,36 @@ void IMDeadCodeElimPass::rewriteModuleSignature(FModuleOp module) {
     // Replace old instance results with dummy wires.
     for (auto index : deadPortIndexes.set_bits()) {
       auto result = instance.getResult(index);
+      assert((!isKnownAlive(result) ||
+              oldPorts[index].direction == Direction::In) &&
+             "If a dead module port is alive in instance results, the "
+             "corresponding port must be input");
+      // If RefType and live, don't want to leave wire around.
+      if (isa<RefType>(result.getType()) && isKnownAlive(result)) {
+        auto getRefDefine = [](Value result) -> RefDefineOp {
+          for (auto *user : result.getUsers()) {
+            if (auto rd = dyn_cast<RefDefineOp>(user);
+                rd && rd.getDest() == result)
+              return rd;
+          }
+          return {};
+        };
+        auto rd = getRefDefine(result);
+        assert(rd && "input ref port to instance is alive, but no driver?");
+        assert(isKnownAlive(rd.getSrc()));
+        result.replaceAllUsesWith(rd.getSrc());
+        ++numErasedOps;
+        rd.erase();
+        continue;
```

Comment:
Maybe it is not assured that `rd.getSrc()` dominates all users of results?

Reply:
Good call, thanks!

For this and other reasons, starting to think need an op to bounce refs through that can be used in these situations -- replacing def of a ref that of course dominates its uses but the uses may not dominate each other (or indeed not all visible at any point).  Something like `%out, %in = firrtl.(ref.)pipe : !firrtl.ref<T>`, or just a wire/wire-like, to plumb dataflow through.  Such an op that guarantees (requires) a single driver is likely useful beyond references.

An op like that would hopefully avoid the need for the special casing done here and in ModuleInliner, or at least reduce it to producing the right sort of bounce declaration.

Looks like IMDCE presently assumes single-block, so one approach would be hoisting backwards_slice(driver), but that doesn't seem like a great approach or something to be bothering about in this pass.

Inliner has similar issues/concerns, and looks like it is intended to work with "when's" -- in some situations if there's input ref (same as the situation here w/IMDCE, for an example of this see [1]) it can produce invalid IR similarly re:dominance.  Both are unreachable presently as we don't generate input reference ports.

[1]  Example input (also crashes this IMDCE for reasons discussed): https://github.com/dtzSiFive/reftype-inputs/blob/ef6294d2c21063f5a26618e7e7b3d1a9c91896c5/inliner-refs.fir .

Reply:
LLVM_DEBUG wrapped dominance check for now?

Reply:
report fatal error if not already in same block and ordered legally.  Shouldn't happen unless we enable input references prematurely or run IMDCE before ExpandWhen's.  I did not want to introduce additional fragility to our passes re:hidden assumptions, but at least this has a guard.  Handling it at all is only needed if there are input references that are used locally but dead through the instance -- which may be good to make illegal anyway.